### PR TITLE
ci(macos): add a macos init/pack smoke test

### LINKED ISSUE #1767: Charmcraft pack does not work when updating to 3.1.1 on osx
### Bug Description

Running `charmcraft pack` fails on a charm that used to build perfectly. I just upgraded from charmcraft 2.6 to 3.1.1 and the error started coming up.
```
cannot inject snap 'charmcraft' from host on a non-linux system
install the snap from the store by setting the 'channel' parameter
```

### To Reproduce

Run `charmcraft pack` on any charm.

### Environment

OSX 14.5

### charmcraft.yaml

```shell
type: charm

# Charm build options
bases:
- name: ubuntu
  channel: "20.04"
  architectures: [amd64]
- name: ubuntu
  channel: "20.04"
  architectures: [arm64]
- name: ubuntu
  channel: "22.04"
  architectures: [amd64]
- name: ubuntu
  channel: "22.04"
  architectures: [arm64]
parts:
  charm:
    build-packages:
    - git
    - libffi-dev
    - libssl-dev
    - rustc
    - cargo
    - pkg-config
```

### Relevant log output

```shell
2024-07-26 16:10:34.135 Starting charmcraft, version 3.1.1
2024-07-26 16:10:34.135 Log verbosity level set to BRIEF
2024-07-26 16:10:34.135 Configuring application...
2024-07-26 16:10:34.135 Preparing application...
2024-07-26 16:10:34.136 Build plan: platform=None, build_for=None
2024-07-26 16:10:34.139 Loading project file '/Users/X/Y/anbox-cloud-charms/charms/gateway-ops/charmcraft.yaml'
2024-07-26 16:10:34.142 Processing grammar (on arm64 for arm64)
2024-07-26 16:10:34.142 Processing grammar for build-packages: ['git', 'libffi-dev', 'libssl-dev', 'rustc', 'cargo', 'pkg-config']
2024-07-26 16:10:34.143 Running charmcraft:ubuntu-20.04-arm64 in arm64 instance...
2024-07-26 16:10:34.143 Preparing managed instance 'charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000'
2024-07-26 16:10:34.144 Using hostname 'charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000'
2024-07-26 16:10:34.144 Not loading snap config because charmcraft is not running as a snap.
2024-07-26 16:10:34.144 No snap config found.
2024-07-26 16:10:34.144 Using default provider 'multipass' on non-linux system.
2024-07-26 16:10:34.144 Executing on host: multipass version
2024-07-26 16:10:34.200 Launching managed ubuntu 20.04 instance...
2024-07-26 16:10:34.200 Executing on host: multipass list --format json
2024-07-26 16:10:34.758 Executing on host: multipass launch snapcraft:20.04 --name charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 --cpus 2 --memory 2G --disk 64G
2024-07-26 16:14:11.751 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- test -f /etc/craft-instance.conf
2024-07-26 16:14:11.886 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:11.951 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.R74rfbhnhq
2024-07-26 16:14:12.010 Created temporary file '/tmp/tmp.R74rfbhnhq' inside instance.
2024-07-26 16:14:12.111 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.R74rfbhnhq
2024-07-26 16:14:12.170 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.R74rfbhnhq
2024-07-26 16:14:12.237 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.R74rfbhnhq /etc/craft-instance.conf
2024-07-26 16:14:12.322 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- cat /etc/os-release
2024-07-26 16:14:12.409 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- test -f /etc/craft-instance.conf
2024-07-26 16:14:12.487 Executing on host: multipass transfer charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000:/etc/craft-instance.conf /Users/X/tmpf9n2wxmm.tmp-craft/tmp331tcifs
2024-07-26 16:14:12.560 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- test -f /etc/craft-instance.conf
2024-07-26 16:14:12.634 Executing on host: multipass transfer charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000:/etc/craft-instance.conf /Users/X/tmpz0lnwyo9.tmp-craft/tmphqkatn20
2024-07-26 16:14:12.705 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:12.777 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.oprXhJECf5
2024-07-26 16:14:12.829 Created temporary file '/tmp/tmp.oprXhJECf5' inside instance.
2024-07-26 16:14:12.893 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.oprXhJECf5
2024-07-26 16:14:12.957 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.oprXhJECf5
2024-07-26 16:14:13.019 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.oprXhJECf5 /etc/craft-instance.conf
2024-07-26 16:14:13.071 No cache path set, not mounting cache directories.
2024-07-26 16:14:13.071 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:13.124 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.kzGXcJXciV
2024-07-26 16:14:13.186 Created temporary file '/tmp/tmp.kzGXcJXciV' inside instance.
2024-07-26 16:14:13.245 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.kzGXcJXciV
2024-07-26 16:14:13.305 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.kzGXcJXciV
2024-07-26 16:14:13.369 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.kzGXcJXciV /etc/environment
2024-07-26 16:14:13.423 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:13.485 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.LAKJISGL1u
2024-07-26 16:14:13.540 Created temporary file '/tmp/tmp.LAKJISGL1u' inside instance.
2024-07-26 16:14:13.608 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.LAKJISGL1u
2024-07-26 16:14:13.670 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.LAKJISGL1u
2024-07-26 16:14:13.733 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.LAKJISGL1u /etc/apt/apt.conf.d/20auto-upgrades
2024-07-26 16:14:13.780 Waiting for environment to be ready...
2024-07-26 16:14:13.780 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl is-system-running
2024-07-26 16:14:13.840 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:13.894 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.A1fOAIxVSW
2024-07-26 16:14:13.958 Created temporary file '/tmp/tmp.A1fOAIxVSW' inside instance.
2024-07-26 16:14:14.022 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.A1fOAIxVSW
2024-07-26 16:14:14.082 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.A1fOAIxVSW
2024-07-26 16:14:14.133 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.A1fOAIxVSW /etc/hostname
2024-07-26 16:14:14.196 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- hostname -F /etc/hostname
2024-07-26 16:14:14.255 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- ln -sf /run/systemd/resolve/resolv.conf /etc/resolv.conf
2024-07-26 16:14:14.322 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl enable systemd-resolved
2024-07-26 16:14:14.536 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl restart systemd-resolved
2024-07-26 16:14:14.656 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:14.732 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.qEjbCHw3eF
2024-07-26 16:14:14.800 Created temporary file '/tmp/tmp.qEjbCHw3eF' inside instance.
2024-07-26 16:14:14.866 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.qEjbCHw3eF
2024-07-26 16:14:14.928 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.qEjbCHw3eF
2024-07-26 16:14:14.986 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.qEjbCHw3eF /etc/systemd/network/10-eth0.network
2024-07-26 16:14:15.036 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl enable systemd-networkd
2024-07-26 16:14:15.239 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl restart systemd-networkd
2024-07-26 16:14:15.415 Waiting for networking to be ready...
2024-07-26 16:14:15.415 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- getent hosts snapcraft.io
2024-07-26 16:14:15.738 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:15.818 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.MlMY3cJFTD
2024-07-26 16:14:15.895 Created temporary file '/tmp/tmp.MlMY3cJFTD' inside instance.
2024-07-26 16:14:15.955 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.MlMY3cJFTD
2024-07-26 16:14:16.017 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.MlMY3cJFTD
2024-07-26 16:14:16.070 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.MlMY3cJFTD /etc/apt/apt.conf.d/00no-recommends
2024-07-26 16:14:16.126 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mktemp
2024-07-26 16:14:16.176 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown ubuntu:ubuntu /tmp/tmp.qOdl4QUBvV
2024-07-26 16:14:16.232 Created temporary file '/tmp/tmp.qOdl4QUBvV' inside instance.
2024-07-26 16:14:16.283 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chown root:root /tmp/tmp.qOdl4QUBvV
2024-07-26 16:14:16.343 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- chmod 0644 /tmp/tmp.qOdl4QUBvV
2024-07-26 16:14:16.404 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- mv /tmp/tmp.qOdl4QUBvV /etc/apt/apt.conf.d/00update-errors
2024-07-26 16:14:16.459 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- apt-get update
2024-07-26 16:14:19.623 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- apt-get install -y apt-utils build-essential curl fuse udev python3 python3-dev python3-pip python3-wheel python3-setuptools gpg dirmngr
2024-07-26 16:15:23.229 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl is-active systemd-udevd
2024-07-26 16:15:23.320 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- apt-get install -y snapd
2024-07-26 16:15:23.572 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- ln -sf /var/lib/snapd/snap /snap
2024-07-26 16:15:23.651 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl enable --now snapd.socket
2024-07-26 16:15:23.822 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- systemctl restart snapd.service
2024-07-26 16:15:24.108 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- snap wait system seed.loaded
2024-07-26 16:15:24.197 Holding refreshes for snaps.
2024-07-26 16:15:24.197 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- snap refresh --hold
2024-07-26 16:15:24.430 Waiting for pending snap refreshes to complete.
2024-07-26 16:15:24.431 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- snap watch '--last=auto-refresh?'
2024-07-26 16:15:24.570 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- snap unset system proxy.http
2024-07-26 16:15:24.780 Executing on host: multipass exec charmcraft-anbox-stream-gateway-on-arm64-for-arm64-58268000 -- sudo -H -- snap unset system proxy.https
2024-07-26 16:15:25.060 Installing snap 'charmcraft' with channel=None and classic=True
2024-07-26 16:15:25.061 cannot inject snap 'charmcraft' from host on a non-linux system
install the snap from the store by setting the 'channel' parameter
2024-07-26 16:15:25.072 Traceback (most recent call last):
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_providers/multipass/multipass_provider.py", line 236, in launched_environment
2024-07-26 16:15:25.072     instance = launch(
2024-07-26 16:15:25.072                ^^^^^^^
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_providers/multipass/_launch.py", line 82, in launch
2024-07-26 16:15:25.072     base_configuration.setup(executor=instance)
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_providers/base.py", line 1044, in setup
2024-07-26 16:15:25.072     self._setup_snaps(executor=executor)
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_providers/base.py", line 917, in _setup_snaps
2024-07-26 16:15:25.072     self._install_snaps(executor=executor)
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_providers/base.py", line 647, in _install_snaps
2024-07-26 16:15:25.072     raise BaseConfigurationError(
2024-07-26 16:15:25.072 craft_providers.errors.BaseConfigurationError: cannot inject snap 'charmcraft' from host on a non-linux system
2024-07-26 16:15:25.072 install the snap from the store by setting the 'channel' parameter
2024-07-26 16:15:25.072 
2024-07-26 16:15:25.072 The above exception was the direct cause of the following exception:
2024-07-26 16:15:25.072 Traceback (most recent call last):
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_application/application.py", line 541, in run
2024-07-26 16:15:25.072     self.run_managed(platform, build_for)
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/charmcraft/application/main.py", line 137, in run_managed
2024-07-26 16:15:25.072     super().run_managed(platform, build_for)
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_application/application.py", line 366, in run_managed
2024-07-26 16:15:25.072     with self.services.provider.instance(
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 137, in __enter__
2024-07-26 16:15:25.072     return next(self.gen)
2024-07-26 16:15:25.072            ^^^^^^^^^^^^^^
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_application/services/provider.py", line 122, in instance
2024-07-26 16:15:25.072     with provider.launched_environment(
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 137, in __enter__
2024-07-26 16:15:25.072     return next(self.gen)
2024-07-26 16:15:25.072            ^^^^^^^^^^^^^^
2024-07-26 16:15:25.072   File "/opt/homebrew/Cellar/charmcraft/3.1.1/libexec/lib/python3.12/site-packages/craft_providers/multipass/multipass_provider.py", line 246, in launched_environment
2024-07-26 16:15:25.072     raise MultipassError(str(error)) from error
2024-07-26 16:15:25.072 craft_providers.multipass.errors.MultipassError: cannot inject snap 'charmcraft' from host on a non-linux system
2024-07-26 16:15:25.072 install the snap from the store by setting the 'channel' parameter
2024-07-26 16:15:25.072 Full execution log: '/Users/X/Library/Logs/charmcraft/charmcraft-20240726-161034.135204.log'
```
```

### PR DESCRIPTION
**Pull request by lengau**
Fixes #1767, depends on #1822

### CODE REVIEW

File: .github/workflows/tests.yaml
Comment (lengau) [inline .github/workflows/tests.yaml:1]: This file contains the only changes in this PR - the rest are in #1822

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -284,3 +284,50 @@ jobs:
...
+          while ! multipass version; do
+            sleep 1
+          done
+      - name: Init and pack
+        run: |
+          mkdir test-charm
+          cd test-charm
+          export CRAFT_VERBOSITY_LEVEL=trace
+          charmcraft init --profile=machine
+          charmcraft pack
```
Comment (tigarmo) [inline .github/workflows/tests.yaml:333]: maybe check for the existence of the packed charm afterwards? just in case

**Review by tigarmo — APPROVED** [review]

**Review by mr-cal — APPROVED** [review]

**Comment by mr-cal** [comment]
Approving put this on the merge queue.  I pulled it off the merge queue since Tiago had a suggestion.

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (lengau) [reply .github/workflows/tests.yaml:333]: Good idea! added.

**Pull request merged**